src/adapters/mongoose.js

```diff
@@ -200,7 +200,15 @@ module.exports = (model, opts) => {
   });
 
   function getRequired(fieldInfo) {
-    return fieldInfo.isRequired === true;
+    return fieldInfo.isRequired === true || fieldInfo.path === '_id';
+  }
+
+  function getGenerated(fieldInfo) {
+    return Boolean(fieldInfo.options && (fieldInfo.options.auto || fieldInfo.options.default));
+  }
+
+  function getReadOnly(fieldInfo) {
+    return fieldInfo.path === '_id';
```

Comment:
Can we have a "readonly" `_id` field that is not "generated"?

Reply:
@arnaudbesnier in mongoose, you can declare an `_id` as a `string`, not an `ObjectId`.

In that case, mongoose will not auto-generate the `_id` and expect you to provide its value when creating a new record.

In that regard, it's possible to have a readonly `_id` that is not auto-generated.